types/parameter/index.d.ts

```diff
@@ -233,7 +234,8 @@ declare namespace Parameter {
         | ParameterRuleObject
         | ParameterRuleArray;
 
-    type ParameterRules = Record<string, ParameterRuleItem | ParameterRuleAbbr | undefined>;
+    type LegalRuleValue = ParameterRuleItem | ParameterRuleAbbr | undefined;
+    type ParameterRules<T extends keyof any = string> = Record<T, LegalRuleValue>;
```

Comment:
Should be implemented using `[key in T]` syntax

Reply:
Okay, I'll look into it~

Reply:
sorry, i don't quite understand whether to use `[key in T]` syntax to check type of index of key? but the type of native object's key is `number` or `string`;

this is the use case of my original design

```typescript
    // type
    type ParameterRule = ParameterRuleItem | ParameterRuleAbbr | undefined;
    type ParameterRules<T extends keyof any = string> = Record<T, ParameterRule>;

    // use case
    interface Params {
        name?: string;
        proejctId: string;
    }

    type ParamsValidator = ParameterRules<keyof Params>;

    interface validateRules:ParamsValidator = {
        name: { type: 'string?', default: '' },
        projectId: 'string',
    };
```

or the `[key in T]` syntax replaces the `Record` Utility Types like this:

```typescript
    // type
    type ParameterRule = ParameterRuleItem | ParameterRuleAbbr | undefined;
    type ParameterRules<T extends keyof any = stirng> = {
        [K in T]: ParameterRule;
    };

    // use case
    interface Params {
        name?: string;
        proejctId: string;
    }

    type ParamsValidator = ParameterRules<keyof Params>;

    interface validateRules:ParamsValidator = {
        name: { type: 'string?', default: '' },
        projectId: 'string',
    };
```

Reply:
Here's the way I suggested:
```
// type
type ParameterRule = 'foo' | 'bar';
type ParameterRules<T = any> = {
  [K in keyof T]?: ParameterRule;
};

// use case
interface Params {
  name?: string;
  projectId: string;
}

const ruleA: ParameterRules<Params> = {
  name: 'foo',
  projectId: 'bar',
};

const ruleB: ParameterRules = {
  foo: 'bar'
};
```

You can try it out in (playground)[https://www.typescriptlang.org/play?ssl=1&ssc=1&pln=20&pc=3#code/PTAEBcE8AcFMCgp1ABQIYCc0FtblhgEoCuANrKALygDkAZgPYM2gA+tARpjQNyIwV0WXPiJlYAZwA8AFSqg0AO0gA+eQG94oUAG0A0qACWi0AGtYkBnVAyAugH4AXKkw48BEuT4BfPvBCgxBIUAMZowfDGonRoIYKu2BKgmtqKbk6gEuAYxgDmfNrQGAwAVrAh4ACSACbOWTmK+fDe8PAhDIpZoBjiAILOQm6inpJSg4lq1CmgabjO9Ew0ADRaoEWl5VW1nNwrvq3tneDd4gBCAwnuYuRJU6uMDPNcGDTNPEA]. Remember to change `ParameterRule` since `'foo' | 'bar'` is just for quick example.

Reply:
nice~ i got it

Reply:
already finish~